jaxopt/_src/lbfgs.py

```diff
@@ -318,11 +319,14 @@ def update(self,
     start = state.iter_num % self.history_size
     value, grad = (state.value, state.grad)
 
-    product = inv_hessian_product(pytree=grad, s_history=state.s_history,
+    if self.history_size:
```

Comment:
Could you explain why we need to fall back like that? At least give a comment with the reason for the change please

Reply:
inv_hessian_product fails for zero history_length. Moved test to inv_hessian.

Reply:
Ok. But was this specific for the complex case or already an issue before? Would you have a test case that was resulting in a bug? Also can't it be handled by the fact that a loop over an empty set just returns the initial value?

Reply:
If I remember correctly, this was already an issue before changing the inverse Hessian loop, LBFGS with history_size=0 gave an error, but possibly at different place.

It seems that when tracing the function the loop is always entered to determine shape, but fails when indexing zero length axes. 

The shortened traceback that I get is:

File ~/Devel/jaxopt/jaxopt/_src/lbfgs.py:41, in select_ith_tree(tree_history, i)
     39 def select_ith_tree(tree_history, i):
     40   """Select tree corresponding to ith history entry"""
---> 41   return tree_map(lambda a: a[i, ...], tree_history)

[...]

IndexError: index is out of bounds for axis 0 with size 0

With this check for history_size==0, LBFG falls back to steepest descent with linesearch, but without scaling of linestep by gamma. It might be useful to keep this gamma scaling, because then stepsize=1 is always (?) a good starting guess. But this would need to always keep one level of history.

Reply:
Ok, great! You're finding many bugs, thank you very much! I'll investigate that in a separate PR like the bug you found for the pytrees. Again, thank you Gregor!